src/dotnet/CommandFactory/CommandResolution/LocalToolsCommandResolver.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Collections.Generic;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.PlatformAbstractions;
+using Microsoft.DotNet.ToolManifest;
+using Microsoft.DotNet.ToolPackage;
+using NuGet.Frameworks;
+using Microsoft.DotNet.Cli;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.CommandFactory
+{
+    internal class LocalToolsCommandResolver : ICommandResolver
+    {
+        private readonly ToolManifestFinder _toolManifest;
+        private readonly ILocalToolsResolverCache _localToolsResolverCache;
+        private readonly IFileSystem _fileSystem;
+        private readonly DirectoryPath _nugetGlobalPackagesFolder;
+        private const string LeadingDotnetPrefix = "dotnet-";
+
+        public LocalToolsCommandResolver(ToolManifestFinder toolManifest = null,
+            ILocalToolsResolverCache localToolsResolverCache = null,
+            IFileSystem fileSystem = null,
+            DirectoryPath? nugetGlobalPackagesFolder = null)
+        {
+            _toolManifest = toolManifest ?? new ToolManifestFinder(new DirectoryPath(Directory.GetCurrentDirectory()));
+            _localToolsResolverCache = localToolsResolverCache ?? new LocalToolsResolverCache();
+            _fileSystem = fileSystem ?? new FileSystemWrapper();
+            _nugetGlobalPackagesFolder =
+                nugetGlobalPackagesFolder ?? new DirectoryPath(NuGetGlobalPackagesFolder.GetLocation());
+        }
+
+        public CommandSpec Resolve(CommandResolverArguments arguments)
+        {
+            if (arguments == null || string.IsNullOrWhiteSpace(arguments.CommandName))
+            {
+                return null;
+            }
+
+            if (!arguments.CommandName.StartsWith(LeadingDotnetPrefix, StringComparison.OrdinalIgnoreCase))
+            {
+                return null;
+            }
+
+            var resolveResultWithoutLeadingDotnet = PackageCommandSpecUsingMuxer(arguments,
+                new ToolCommandName(arguments.CommandName.Replace(LeadingDotnetPrefix, string.Empty,
```

Comment:
Should this be a `Substring` from the length of the prefix? As silly as it might be, I would expected a command name of `dotnet-dotnet-foo` to work as `dotnet dotnet-foo`, but this `Replace` call would change it to just `foo`.

Reply:
Right, but that doesn't preclude a command name of `dotnet-dotnet-foo`, since that does start with `dotnet-` and then the `Replace` call removes both instead of just the first prefix.  It's obviously a total corner case, but if we want consistency for all names, I'd suggest using `Substring` over `Replace`.

Reply:
Ah, i see, yes, i should use substring here

Reply:
Fixed, used substring instead